planetiler-core/src/main/java/com/onthegomap/planetiler/pmtiles/Pmtiles.java

```diff
@@ -0,0 +1,332 @@
+
+      try {
+        buffer.get(magic);
+        if (!(new String(magic, StandardCharsets.UTF_8).equals("PMTiles"))) {
+          throw new FileFormatException("Incorrect magic number for PMTiles archive.");
+        }
+
+        byte specVersion = buffer.get();
+        long rootDirOffset = buffer.getLong();
+        long rootDirLength = buffer.getLong();
+        long jsonMetadataOffset = buffer.getLong();
+        long jsonMetadataLength = buffer.getLong();
+        long leafDirectoriesOffset = buffer.getLong();
+        long leafDirectoriesLength = buffer.getLong();
+        long tileDataOffset = buffer.getLong();
+        long tileDataLength = buffer.getLong();
+        long numAddressedTiles = buffer.getLong();
+        long numTileEntries = buffer.getLong();
+        long numTileContents = buffer.getLong();
+        boolean clustered = (buffer.get() == 0x1);
+        Compression internalCompression = Compression.fromByte(buffer.get());
+        Compression tileCompression = Compression.fromByte(buffer.get());
+        TileType tileType = TileType.fromByte(buffer.get());
+        byte minZoom = buffer.get();
+        byte maxZoom = buffer.get();
+        int minLonE7 = buffer.getInt();
+        int minLatE7 = buffer.getInt();
+        int maxLonE7 = buffer.getInt();
+        int maxLatE7 = buffer.getInt();
+        byte centerZoom = buffer.get();
+        int centerLonE7 = buffer.getInt();
+        int centerLatE7 = buffer.getInt();
+        return new Header(
+          specVersion,
+          rootDirOffset,
+          rootDirLength,
+          jsonMetadataOffset,
+          jsonMetadataLength,
+          leafDirectoriesOffset,
+          leafDirectoriesLength,
+          tileDataOffset,
+          tileDataLength,
+          numAddressedTiles,
+          numTileEntries,
+          numTileContents,
+          clustered,
+          internalCompression,
+          tileCompression,
+          tileType,
+          minZoom,
+          maxZoom,
+          minLonE7,
+          minLatE7,
+          maxLonE7,
+          maxLatE7,
+          centerZoom,
+          centerLonE7,
+          centerLatE7
+        );
+      } catch (BufferUnderflowException e) {
+        throw new FileFormatException("Failed to read enough bytes for PMTiles header.");
+      }
+    }
+  }
+
+  /**
+   * Convert a directory of entries to bytes.
+   *
+   * @param slice a list of entries sorted by ascending {@code tileId} with size > 0.
+   * @param start the start index to serialize, inclusive.
+   * @param end   the end index, exclusive.
+   * @return the uncompressed bytes of the directory.
+   */
+  public static byte[] serializeDirectory(List<Entry> slice, int start, int end) {
+    ByteArrayList dir = new ByteArrayList();
+
+    VarInt.putVarLong((long) end - start, dir);
+
+    long lastId = 0;
+    for (int i = start; i < end; i++) {
+      VarInt.putVarLong(slice.get(i).tileId - lastId, dir);
+      lastId = slice.get(i).tileId;
+    }
```

Comment:
Since `slice` could be any kind of list here (i.e. linked list), the only other possible tweak here would be to extract a `subList` if start/end aren't the full size of slice, then iterate through it for each of these loops, something like:

```suggestion
    List<Entry> range = start == 0 && end == slice.size() ? slice : slice.subList(start, end);
    long lastId = 0;
    for (var entry : range) {
      VarInt.putVarLong(range.tileId - lastId, dir);
      lastId = range.tileId;
    }
```

That should make it efficient for any list implementation - although a non-arraylist is unlikely.

Reply:
Also, should there be a `serializeDirectory(slice)` convenience method that does the entire list? the range one could construct the sublist and pass it in to avoid the conditional check on start/end size?